### PR TITLE
Discovery search handler at hnsw level

### PR DESCRIPTION
**Pull request by coszio**
Depends on #2834 and #2820. Tracked by #2790 

Handles the case of `QueryVector::Discovery(_)` at the segment level, which consists of two parts:
1. Do a Context search to get best entrypoints
2. Use the entrypoints to do the actual Discovery search

## TODO

- [x] tests

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
4. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
5. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/segment/src/index/hnsw_index/hnsw.rs [hunk 1]
```diff
@@ -268,7 +269,16 @@ impl<TGraphLinks: GraphLinks> HNSWIndex<TGraphLinks> {
...
+            .map(|&vector| match vector {
+                QueryVector::Discovery(discovery_query) => self.discovery_search_with_graph(
+                    discovery_query.clone(),
+                    filter,
+                    top,
+                    params,
+                    is_stopped,
+                ),
+                other => self.search_with_graph(other, filter, top, params, None, is_stopped),
+            })
```
Comment (coszio) [inline lib/segment/src/index/hnsw_index/hnsw.rs:281]: Main change

File: lib/segment/tests/integration/filtrable_hnsw_test.rs [hunk 2]
```diff
@@ -16,11 +17,64 @@ use segment::types::{
...
+    match variant {
+        QueryVariant::Nearest => random_vector(rnd, dim).into(),
+        QueryVariant::Discovery => random_discovery_query(rnd, dim),
+        QueryVariant::RecommendBestScore => random_reco_query(rnd, dim),
+    }
+}
+
+#[rstest]
+#[case::nearest(QueryVariant::Nearest, 5)]
+#[case::discovery(QueryVariant::Discovery, 25)]
```
Comment (coszio) [inline lib/segment/tests/integration/filtrable_hnsw_test.rs:72]: This level of precision is not great, but it is what's needed to pass the test. 

Something curious: I experimented with and without the 2-step search, and got indistinguishable changes in precision here.

Comment (generall) [reply lib/segment/tests/integration/filtrable_hnsw_test.rs:72]: tbh, I am a bit concerned with this numbers, maybe we would need to investigate a bit deeper

File: lib/segment/src/index/hnsw_index/hnsw.rs [hunk 3]
```diff
@@ -668,16 +717,22 @@ impl<TGraphLinks: GraphLinks> VectorIndex for HNSWIndex<TGraphLinks> {
...
 
                 // It is expected, that graph will become disconnected less than
                 // $1/m$ points left.
                 // So blocks larger than $1/m$ are not needed.
                 // We add multiplier for the extra safety.
-                let percolation_multiplier = 2;
+                let percolation_multiplier = 4;
                 let max_block_size = if self.config.m > 0 {
-                    total_vector_count / self.config.m * percolation_multiplier
+                    total_vector_count / average_links_per_0_level_int * percolation_multiplier
```
Comment (generall) [inline lib/segment/src/index/hnsw_index/hnsw.rs:735]: This pair of changes is not directly related to the PR, but it helps to test changes more easily, so I decided to include them anyway.
The original problem is, that apparently current thresholds are too weak to address precision loss in out test dataset, so this changes here should improve performance in similar cases. To better understand which values are the most optimal, some additional experiments should be conducted.


The first change, `percolation_multiplier` should allow larger payload blocks to create an additional links in HNSW graph. Original multiplier value was selected randomly, so changing it to another arbitrary values, even with just anecdotal evidence of effectiveness, should be OK.

Second change  - `average_links_per_0_level_int` improves threshold prediction by measuring actual number of average connection per node, which might be less than expected due to HNSW heuristic that we use.

**Review by generall — APPROVED** [review]

**Pull request merged**